### PR TITLE
Rework RedirectProccess to have Builder API

### PR DESCRIPTION
**Pull request by gkassabli**
#### Description
Following the https://github.com/spatialos/gdk-for-unity/pull/670/files I've tried the idea to rework RedirectedProcess to have Builder API

#### Tests
Ensure the code using RedirectedProcess runs the same

### CODE REVIEW

File: workers/unity/Packages/com.improbable.gdk.buildsystem/WorkerBuilder.cs [hunk 1]
```diff
@@ -257,10 +257,11 @@ private static void Zip(string zipAbsolutePath, string basePath, bool useCompres
         {
             using (new ShowProgressBarScope($"Package {basePath}"))
             {
-                RedirectedProcess.Run(Common.SpatialBinary, "file", "zip",
+                RedirectedProcess.CommandWithArgs(Common.SpatialBinary, "file", "zip",
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.buildsystem/WorkerBuilder.cs]: I'd prefer an explicit:

```
RedirectedProgress.Command(Common.SpatialBinary)
	.WithArgs(...)
```

File: workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs [hunk 2]
```diff
@@ -23,15 +23,46 @@ public class RedirectedProcessResult
     /// </summary>
     public class RedirectedProcess
     {
+        private string RunningCommand = string.Empty;
+        private string[] Arguments;
+        private string WorkingDirectory = Path.GetFullPath(Path.Combine(Application.dataPath, ".."));
+        private Action<string> OutputProcessor;
+        private Action<string> ErrorProcessor;
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs]: `private` variables should be camel case.

File: workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs [hunk 3]
```diff
@@ -23,15 +23,46 @@ public class RedirectedProcessResult
...
         ///     Runs the redirected process and waits for it to return.
         /// </summary>
         /// <param name="command">The filename to run.</param>
         /// <param name="arguments">Parameters that will be passed to the command.</param>
         /// <returns>The exit code.</returns>
-        public static int Run(string command, params string[] arguments)
+        // public static int Run(string command, params string[] arguments)
+        // {
+        //     return RunIn(Path.GetFullPath(Path.Combine(Application.dataPath, "..")), command, arguments);
+        // }
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs]: Commented code.

File: workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs [hunk 4]
```diff
@@ -64,21 +96,35 @@ public static int RunIn(string workingDirectory, string command, params string[]
 
                 var processOutput = new StringBuilder();
 
-                void OnReceived(object sender, DataReceivedEventArgs args)
+                void OnReceived(string output)
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs]: These probably shouldn't be local methods here. As we allocate the string builder regardless of whether we end up using this

File: workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs [hunk 5]
```diff
@@ -107,83 +168,49 @@ void OnReceived(object sender, DataReceivedEventArgs args)
...
         ///     Runs the redirected process and returns a task which can be waited on.
         /// </summary>
-        /// <param name="workingDirectory">The directory to run the command from.</param>
-        /// <param name="command">The filename to run.</param>
-        /// <param name="arguments">Parameters that will be passed to the command.</param>
         /// <param name="redirectStdout">Redirect standard output to Debug.Log</param>
         /// <param name="redirectStderr">Redirect standard error to Debug.LogError</param>
         /// <returns>A task which would return the exit code and output.</returns>
-        public static async Task<RedirectedProcessResult> RunInAsync(string workingDirectory, string command, string[] arguments, bool redirectStdout = false, bool redirectStderr = false)
+        public async Task<RedirectedProcessResult> RunInAsync(bool redirectStdout = false, bool redirectStderr = false)
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs]: Shouldn't this be RunAsync and the In part of it is folded back into whether or not that option has been set?

Comment (gkassabli) [reply workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs]: Kept the name as is - I'll rename it

File: workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs [hunk 6]
```diff
@@ -107,83 +168,49 @@ void OnReceived(object sender, DataReceivedEventArgs args)
...
         ///     Runs the redirected process and returns a task which can be waited on.
         /// </summary>
-        /// <param name="workingDirectory">The directory to run the command from.</param>
-        /// <param name="command">The filename to run.</param>
-        /// <param name="arguments">Parameters that will be passed to the command.</param>
         /// <param name="redirectStdout">Redirect standard output to Debug.Log</param>
         /// <param name="redirectStderr">Redirect standard error to Debug.LogError</param>
         /// <returns>A task which would return the exit code and output.</returns>
-        public static async Task<RedirectedProcessResult> RunInAsync(string workingDirectory, string command, string[] arguments, bool redirectStdout = false, bool redirectStderr = false)
+        public async Task<RedirectedProcessResult> RunAsync(bool redirectStdout = false, bool redirectStderr = false)
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs]: `bool`-based APIs generally smell a bit. Prefer an `enum` as flags here instead. 
For example, `[Flags] enum { RedirectStdErr, RedirectStdOut, RedirectAll }`
This avoid call sites that look like `RunAsync(true, true) <-- huh?` or the verbose `RunAsync(redirectStdout: true, redirectStderr: true)`.
Also, as a general point, `bools` tend to multiply and encourage "customizing" functions that should really be split out into their own separate area of concern. I've seen monstrosities like `thing.DoUpdate(false, 7, true, true)` more times than I can count.

File: workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs [hunk 7]
```diff
@@ -107,83 +168,49 @@ void OnReceived(object sender, DataReceivedEventArgs args)
...
-            var info = new ProcessStartInfo(command, string.Join(" ", arguments))
-            {
-                CreateNoWindow = true,
-                RedirectStandardError = true,
-                RedirectStandardOutput = true,
-                UseShellExecute = false,
-                WorkingDirectory = workingDirectory
-            };
-
             return await Task.Run(() =>
```
[the remark this answers is not in the document]
Comment (gkassabli) [reply workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs]: Left this part as is. cc @davedissian

File: workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs (hunk 6 above)
Comment (gkassabli) [reply workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs]: Good point, let me rework this API

File: workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs [hunk 8]
```diff
@@ -17,31 +17,114 @@ public class RedirectedProcessResult
         public List<string> Stderr;
     }
 
+    [Flags]
+    public enum ProcessOutputOptions
+    {
+        /// <summary>
+        ///   <para>No redirected output, only custom outputProcessors are used</para>
+        /// </summary>
+        None = 0,
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs]: `NoRedirect` ?

File: workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs [hunk 9]
```diff
@@ -60,44 +143,101 @@ public static int RunIn(string workingDirectory, string command, params string[]
...
+                {
+                    return process.ExitCode;
+                }
+
                 // Ensure that the first line of the log is something useful in the Unity editor console.
-                var trimmedOutput = processOutput.ToString().Trim();
+                var trimmedOutput = outputLog.ToString().TrimStart();
 
-                if (!string.IsNullOrEmpty(trimmedOutput))
+                if (string.IsNullOrEmpty(trimmedOutput))
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs]: Nit - Only need to check for `empty` here

**Comment by jamiebrynes7** [comment]
Oh also! Changelog 😉 

Its technically a public API that someone may be using.

File: workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs (hunk 8 above)
Comment (gkassabli) [reply workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs]: This is set of flags, not direct enum, so `.NoRedirect | .RedirectStdOut` is a valid assignment, though doesn't make a lot of sense semantically. None also doesn't look great in this case, but I am not sure what would be a good naming

Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs]: Alternatively you could rename the enum to `RedirectBehaviour` in which case `None` makes more sense 💯

File: workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jared-improbable) [reply workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs]: Then you definitely want to remove the `await` here

**Pull request merged**